be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp

```diff
@@ -883,45 +951,160 @@ Status VerticalSegmentWriter::_merge_rows_for_sequence_column(
+        const std::vector<vectorized::IOlapColumnDataAccessor*>& key_columns,
+        vectorized::IOlapColumnDataAccessor* seq_column, const signed char* delete_signs,
+        const std::vector<RowsetSharedPtr>& specified_rowsets,
+        std::vector<std::unique_ptr<SegmentCacheHandle>>& segment_caches) {
+    VLOG_DEBUG << fmt::format(
+            "VerticalSegmentWriter::_merge_rows_for_insert_after_delete enter: data.block:{}\n",
+            data.block->dump_data());
+    // there will be at most 2 rows for a specified key in block when control flow reaches here
+    // after this function, there will not be duplicate rows in block
+    auto filter_column = vectorized::ColumnUInt8::create(data.num_rows, 1);
+    auto* __restrict filter_map = filter_column->get_data().data();
+    std::string previous_key {};
+    bool previous_has_delete_sign {false};
+    int duplicate_rows {0};
+    int32_t delete_sign_col_unique_id =
+            _tablet_schema->column(_tablet_schema->delete_sign_idx()).unique_id();
+    for (size_t block_pos = data.row_pos; block_pos < data.row_pos + data.num_rows; block_pos++) {
+        size_t delta_pos = block_pos - data.row_pos;
+        auto& skip_bitmap = skip_bitmaps->at(block_pos);
+        std::string key = _full_encode_keys(key_columns, delta_pos);
+        bool have_delete_sign =
+                (!skip_bitmap.contains(delete_sign_col_unique_id) && delete_signs[block_pos] != 0);
+        if (delta_pos > 0 && previous_key == key) {
+            // !!ATTENTION!!: We can only remove the row with delete sign if there is a insert with the same key after this row.
+            // If there is only a row with delete sign, we should keep it and can't remove it from block, because
+            // compaction will not use the delete bitmap when reading data. So there may still be rows with delete sign
+            // in later process
+            DCHECK(previous_has_delete_sign);
+            DCHECK(!have_delete_sign);
+            ++duplicate_rows;
+            RowLocation loc;
+            RowsetSharedPtr rowset;
+            Status st = _tablet->lookup_row_key(key, _tablet_schema.get(), false, specified_rowsets,
+                                                &loc, _mow_context->max_version, segment_caches,
+                                                &rowset, true);
+            DCHECK(st.is<KEY_NOT_FOUND>() || st.ok());
+
+            Slice previous_seq_slice {};
+            if (st.ok()) {
+                // delete the existing row
+                _mow_context->delete_bitmap->add(
+                        {loc.rowset_id, loc.segment_id, DeleteBitmap::TEMP_VERSION_COMMON},
+                        loc.row_id);
+                VLOG_DEBUG << fmt::format(
+                        "_merge_rows_for_insert_after_delete: delete existing row, rowset_id={}, "
+                        "segment_id={}, row_id={}",
+                        loc.rowset_id.to_string(), loc.segment_id, loc.row_id);
```

Comment:
Maybe we should implement a new VerticalSegmentWriter inherts from VerticalSegmentWriter which allows duplicated rows.